lib/propolis/src/block/attachment.rs

```diff
@@ -0,0 +1,362 @@
+            Some(Err(e)) => Err(e),
+            Some(Ok(r)) => Ok(r),
+        }
+    }
+
+    /// Block (synchronously) in order to retrieve the next [`Request`] from the
+    /// device.  Will return [`None`] if no device is attached, or the backend
+    /// is stopped, otherwise it will block until a request is available.
+    pub fn block_for_req(&self) -> Option<Request> {
+        match self.0.access(|data, state| {
+            // If we do not immediately get a request, clone the BackendData
+            // so we can do our blocking through that, instead of repeated calls
+            // into BackAttachment::access()
+            data.next_req(state).map_err(|_e| (data.clone(), state.clone()))
+        })? {
+            Ok(req) => Some(req),
+            Err((be, state)) => be.block_for_req(&state),
+        }
+    }
+
+    /// Get a [Waiter], through which one can asynchronously poll for new
+    /// requests via [`Waiter::for_req()`].  Will return [`None`] if no device
+    /// is attached, or the backend is stopped.
+    pub fn waiter(&self) -> Option<Waiter> {
+        self.0.access(|data, state| Waiter(data.clone(), state.clone()))
+    }
+
+    /// Run provided function against [`MemAccessor`] for this backend.
+    ///
+    /// Intended to provide caller with means of creating/associated child
+    /// accessors.
+    pub fn accessor_mem<R>(
+        &self,
+        f: impl FnOnce(&MemAccessor) -> R,
+    ) -> Option<R> {
+        self.0.access(|data, _state| f(&data.acc_mem))
+    }
+
+    /// Assert stopped state on this Attachment
+    pub fn stop(&self) {
+        self.0.access_fallback(
+            |data, state| {
+                state.stop();
+                data.notify();
+            },
+            |state| state.stop(),
+        );
+    }
+
+    /// Clear stopped state from this Attachment
+    pub fn start(&self) {
+        self.0.access_fallback(
+            |data, state| {
+                state.start();
+                data.notify();
+            },
+            |state| state.start(),
+        );
+    }
+
+    /// Attempt to detach backend from device
+    pub fn detach(&self) -> Result<(), attachment::DetachError> {
+        if let Some((dev, backend)) = self
+            .0
+            .access(|data, _state| (data.device.clone(), data.backend.clone()))
+        {
+            detach(dev, backend)
+        } else {
+            Err(attachment::DetachError::BackNotAttached)
+        }
+    }
+}
+
+pub struct Waiter(Arc<BlockData>, attachment::AttachState);
+impl Waiter {
+    /// Wait (via a [`Future`]) to retrieve the next [`Request`] from the
+    /// device.
+    pub fn for_req(&self) -> WaitForReq<'_> {
+        WaitForReq { be: &self.0, att_state: &self.1, wait: self.1.notified() }
+    }
+}
+
+pin_project! {
+    /// Future returned from [`Waiter::wait_for_req()`]
+    ///
+    /// This future is not fused, so it can be repeatedly polled for additional
+    /// [`Request`]s as they become available.
```

Comment:
No existing consumer depends on this style of behavior (subsequent calls to `WaitForReq::poll()` after receiving a `Ready`), so I'll just remove the note for now.  Turning this into a `Stream` would complicate existing consumers, since they would then need to take care of pinning the future themselves (since the pin projection makes `WaitForReq` `!Unpin`)

If a future consumer wanted `Stream` behavior, it wouldn't be hard to impl the trait as a call to this unfused `poll()`